### PR TITLE
LTD-1368: Update Party model to include new fields for End-user

### PR DESCRIPTION
**Pull request by saruniitr**
## Change description

In the journey of adding an End-user we are capturing new details.
We are asking whether there is an end user document available or not, if it is not in english and that atleast one of the page in the document is on company letterhead etc. These changes are specific to End-user only and not applicable to other parties such as Consignee, third parties etc

Introduce party document type to differentiate the documents uploaded by various parties. End-user can upload multiple documents. Add other methods to edit party details, delete a specific party document.
Existing method to delete party document deletes all documents (as it assumes they all upload one document)

### CODE REVIEW

File: api/applications/libraries/document_helpers.py [hunk 1]
```diff
@@ -96,17 +97,20 @@ def upload_party_document(party, data, application, user):
         return JsonResponse(data={"error": "No such user"}, status=status.HTTP_404_NOT_FOUND)
 
     documents = PartyDocument.objects.filter(party=party, safe=True)
-    if documents.exists():
+    if party.type != PartyType.END_USER and documents.exists():
         return JsonResponse(data={"error": "Document already exists"}, status=status.HTTP_400_BAD_REQUEST)
 
     data["party"] = party.id
+    if not data.get("type"):
```
Comment (saruniitr) [inline api/applications/libraries/document_helpers.py:104]: documents uploaded by parties other than End-user are considered supporting documents. Those forms don't supply the type so we consider them as supporting documents.

File: api/applications/views/party_documents.py [hunk 2]
```diff
@@ -34,3 +39,30 @@ def delete(self, request, pk, party_pk):
         application = get_application(pk)
         party = application.get_party(party_pk)
         return delete_party_document(party, application, request.user)
+
+    @transaction.atomic
+    @authorised_to_view_application(ExporterUser)
+    def delete(self, request, pk, party_pk, document_pk):
```
Comment (saruniitr) [inline api/applications/views/party_documents.py:45]: This deletes a specific document as the above method deletes all documents of a `Party`

File: api/parties/enums.py [hunk 3]
```diff
@@ -65,3 +65,17 @@ class PartyRole:
         (CUSTOMER, "Customer"),
         (OTHER, "Other"),
     ]
+
+
+class PartyDocumentType:
+    SUPPORTING_DOCUMENT = "supporting_document"
+    END_USER_UNDERTAKING_DOCUMENT = "end_user_undertaking_document"
+    END_USER_ENGLISH_TRANSLATION_DOCUMENT = "end_user_english_translation_document"
+    END_USER_COMPANY_LETTERHEAD_DOCUMENT = "end_user_company_letterhead_document"
```
Comment (saruniitr) [inline api/parties/enums.py:74]: End-user can upload three different documents so differentiate them this document type is introduced. There is no technical name for the documents uploaded by other parties so they are considered as supporting documents.

**Review by kevincarrogan — COMMENTED** [review]
Can the CI issues be fixed?

File: api/applications/views/parties.py [hunk 4]
```diff
@@ -119,6 +116,28 @@ def get(self, request, pk):
 
         return JsonResponse(data={key: parties_data})
 
+    @authorised_to_view_application(ExporterUser)
+    def get(self, request, pk, party_pk):
+        """
+        Get party detail
+        """
+        application = get_application(pk)
+        party = application.active_parties.get(party_id=party_pk)
```
Comment (kevincarrogan) [inline api/applications/views/parties.py:125]: This appears to be done in both `get` and `post` so maybe better to extract this as a property or a method.

File: api/applications/libraries/document_helpers.py [hunk 5]
```diff
@@ -96,17 +97,20 @@ def upload_party_document(party, data, application, user):
...
+
     serializer = PartyDocumentSerializer(data=data)
 
     if not serializer.is_valid():
         return JsonResponse({"errors": serializer.errors}, status=status.HTTP_400_BAD_REQUEST)
 
-    # Delete existing document with a virus if one exists
-    PartyDocument.objects.filter(party=party, safe=False).delete()
+    # Delete existing document if one exists
+    PartyDocument.objects.filter(party=party, type=data["type"]).delete()
```
[the remark this answers is not in the document]
Comment (saruniitr) [reply api/applications/libraries/document_helpers.py:113]: End-user can upload multiple documents (undertaking, english translation, document on letterhead) and from summary page we have the ability to edit these fields so he can reupload a specific document hence modified this to delete all previous versions of a document of specific type.

File: api/applications/views/parties.py (hunk 4 above)
Comment (saruniitr) [reply api/applications/views/parties.py:125]: :+1:

File: api/applications/views/party_documents.py [hunk 6]
```diff
@@ -34,3 +39,30 @@ def delete(self, request, pk, party_pk):
...
+
+            audit_trail_service.create(
+                actor=request.user,
+                verb=AuditType.DELETE_PARTY_DOCUMENT,
+                target=application.get_case(),
+                payload={"party_type": party.type.replace("_", " "), "party_name": party.name, "file_name": document.name},
+            )
+
+            return HttpResponse(status=status.HTTP_204_NO_CONTENT)
+        except PartyDocument.DoesNotExist:
```
[the remark this answers is not in the document]
Comment (saruniitr) [reply api/applications/views/party_documents.py:64]: :+1:

File: api/applications/views/parties.py [hunk 7]
```diff
@@ -102,13 +112,18 @@ def delete(self, request, pk, party_pk):
...
         Get parties for an application
         """
-        application = get_application(pk)
+        party_pk = self.kwargs.get("party_pk")
+        if party_pk:
+            party_data = PartySerializer(self.party).data
+            return JsonResponse(data={"type": party_data["type"], "data": party_data})
 
-        application_parties = application.active_parties.all().filter(deleted_at__isnull=True).select_related("party")
+        application_parties = (
```
Comment (saruniitr) [inline api/applications/views/parties.py:124]: from here it is existing implementation which returns all parties

File: api/applications/views/parties.py [hunk 8]
```diff
@@ -102,13 +112,18 @@ def delete(self, request, pk, party_pk):
         return JsonResponse(data={"party": PartySerializer(poa.party).data}, status=status.HTTP_200_OK)
 
     @authorised_to_view_application(ExporterUser)
-    def get(self, request, pk):
+    def get(self, request, **kwargs):
         """
         Get parties for an application
         """
-        application = get_application(pk)
+        party_pk = self.kwargs.get("party_pk")
```
Comment (saruniitr) [inline api/applications/views/parties.py:119]: this just returns single `Party` detail. Initially added a separate `get` method but linter was complaining about redefinition so modified it like this.

File: api/applications/views/party_documents.py [hunk 9]
```diff
@@ -15,22 +20,47 @@ class PartyDocumentView(APIView):
...
     @transaction.atomic
     @authorised_to_view_application(ExporterUser)
-    def delete(self, request, pk, party_pk):
-        application = get_application(pk)
-        party = application.get_party(party_pk)
-        return delete_party_document(party, application, request.user)
+    def delete(self, request, **kwargs):
+        document_pk = self.kwargs.get("document_pk")
+        if not document_pk:
+            return delete_party_document(self.party, self.application, request.user)
```
Comment (saruniitr) [inline api/applications/views/party_documents.py:45]: this is existing implementation which assumes only one document for each party but with these new changes end-user can upload multiple documents so we need to be able to delete a specific document.

I have added it as a separate method but linter was complaining about redefinition of `delete` method.

**Review by kevincarrogan — APPROVED** [review]

**Pull request merged**